### ISSUE TITLE
GEP: Add Name to HTTPRouteRule and HTTPRouteMatch

### ISSUE THREAD

**Issue by hzxuzhonghu** [kubernetes-sigs/gateway-api#995]
**What would you like to be added**:

Add a name field to HTTPRouteRule and HTTPRouteMatch structs to allow users to identify different routes 

**Why this is needed**:

- Policy Attachment

- Route Attachment/Delegation: Ability to have more fine grained matches for policy references IF we add sectionName to the policy object

- Improved Debugging: Ability to add debugging information, such as adding the route matched to some access log

**Comment by shaneutt** [comment]
If I'm understanding correctly based on the closed PR, we've decided not to move forward with this at this time. If that's not the case, please feel free to drop a comment and we can re-open.

**Comment by robscott** [comment]
I think we decided this would be reasonable if we had a bigger use case associated with it. This will likely end up being included as part of route delegation in #1058.

**Comment by shaneutt** [comment]
Ok. The project state also says "abandoned" so we should probably update this too so this issue doesn't look as though it is in fact abandoned.

**Comment by shaneutt** [comment]
Ok, then let's consider this closed in favor of the implementation in https://github.com/kubernetes-sigs/gateway-api/issues/1058 :+1:

**Comment by youngnick** [comment]
I think it makes sense in general, but will need careful writing of the spec - we can't add the new field as required, but what value should  be used in contexts where it's expected (like log messages or telemetry) if it's not set?

**Comment by howardjohn** [comment]
`""`, `"-"`, `"unknown"`, `nil` - not sure. It may be usage specific, some systems have different ways to describe "not present" I would think, which makes it hard to prescribe against them

**Comment by swathinsankaran** [comment]
@youngnick @howardjohn We would really appreciate having this in the next version of the HTTPRoute. We have a similar requirement to apply some policies per rule as mentioned in the description.

cc: @apalsule

**Comment by maleck13** [comment]
Within the Kuadrant project, we also see value in adding this to support policy. Would really like to see it in the next version also

**Comment by robscott** [comment]
One of the most common arguments against this feature is that we really shouldn't be encouraging users to have large Routes. Routes with lots of rules are remarkably hard to deal with in cases like partially invalid config where many Route rules are valid, but one is not. We still have not solved this problem so the API only provides guidance for when a Route is invalid.

It seems like any request for attaching policies and other resources to a specific Route Rule name has to answer this key question: Why not just use more than one Route? Any Route that has 10 rules could also be equally represented by 10 Routes with 1 rule each, or 5 Routes with 2 rules each, or 2 Routes with 5 rules each, etc. I'm concerned that adding name to Route rules would encourage even larger Routes, which would only exacerbate other problems that we have not solved yet like partially valid Routes.

**Comment by howardjohn** [comment]
@robscott I think everything you said is true, but also if you have >1 route having a name is useful. So even for a small 2 route service its helpful. 

If its such a concern, why do we even allow multiple routes at all? 

Its very likely too late to change that so mostly a theoretical argument. But it does seem compelling that if you have multiple things, you ought to have a way to refer to them. Listener names, service port names, container names, etc all follow this pattern.

**Comment by youngnick** [comment]
We currently have a maximum items set of 16 in the validation for rules across all the rule types, so I think we're managing the larger routes concern to some extent. I think I should be clear that we will be _very_ reluctant to allow more than that (I actually think 16 is probably too large as it is, but we can't make it smaller without an API version bump).

That said, I think I'm in agreement with @howardjohn here - it is useful for things to have a `name` field if there's a list.

However, versioning means that we have the following things to address in this proposal:
- The new `name` field _must_ be optional, or it's a breaking change.
- Because the field must be optional, we need to define cases like:
  - What implementations should do with empty `name` fields for logging. Leave the field out, leave it empty? Doesn't really matter what we choose, we just need to choose something.
  - What happens when something (like a Policy or whatever) refers to a `name` field that doesn't exist? To me, this is a ResolvedRefs error, the same as if the object as a whole did not exist. But again, we need an answer.

As long as we carefully handle the implications of this field being optional, I think it's okay to proceed.

**Comment by maleck13** [comment]
My 2 cents:
> The new name field must be optional, or it's a breaking change.

I think this is unfortunate but understandable.

> What implementations should do with empty name fields for logging. Leave the field out, leave it empty? Doesn't really matter what we choose, we just need to choose something.

Leave it empty would be my preference.

> What implementations should do with empty name fields for logging. Leave the field out, leave it empty? Doesn't really matter what we choose, we just need to choose something.

> What happens when something (like a Policy or whatever) refers to a name field that doesn't exist? To me, this is a ResolvedRefs error, the same as if the object as a whole did not exist. But again, we need an answer.

Agree again. As I have a particular interest in policy, for me the name should likely become part of the target ref as called out in the future work section of the policy attachment GEP. For me the all of the target ref in a policy has to be validated before the policy is applied. So it would be invalid to have a policy with a target ref that matched the resource but had an invalid `sectionName` ref. 

My 2 cents

**Comment by pleshakov** [comment]
@swathinsankaran 

> We have a similar requirement to apply some policies per rule as mentioned in the description.

Could this potentially be done by using a filter referenced from a routing rule?

**Comment by swathinsankaran** [comment]
@pleshakov Yes, what we are looking for is this [GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713/) to attach the _policy_ per _HTTPRouteRule_ and may be _policy_ per backendRef (like policy to reencrypt traffic).
I have observed that GEP is not progressing, and this issue being a subset of the GEP, will help us going forward with our implementation.

**Comment by maleck13** [comment]
@pleshakov A filter currently cannot be provided by a 3rd party. With kuadrant we are building out a set of policy APIs that hope to provide to many gateway providers. Policy Attachment allows this type of set up, whereas a filter has to be provided by the gateway provider AFAIK

**Comment by youngnick** [comment]
Right now, there's #2128 to add some clarify to Policy Attachment's discoverability, but I agree that this would be nice to address soon.

#2128 is also going to have a bunch of other followup actions, so this issue will be in good company.

Also, I don't think we explicitly disallow third party filters right now, we just had never considered them - I suspect that they will be heavily dependent on the underlying dataplane though, as the Gateway implementation will need to be able to direct its underlying dataplane to send traffic to a third-party filter somehow. That would need configuration somehow, which we also haven't though about yet.

Hmm, reading over that last paragraph, I agree Policy Attachment is probably easier. 😄 

All of that said, @swathinsankaran, I'm going to remove the status from this, and we can figure out if it needs to be in 1.0 or not.

/remove-lifecycle rotten
/lifecycle frozen

**Comment by maleck13** [comment]
@youngnick Wondering if this has been discussed further for 1.0? If there is anything I can do to help progress it let me know

**Comment by shaneutt** [comment]
We triaged this for GA and we think it's indeed not something that needs to block GA, but also we don't think it's actually ready for someone to pick up:

/label-remove help

We expect more conversation and considerations are needed before we can start taking action on this one:

/triage needs-information

**Comment by guicassolato** [comment]
It turns out not having the name field available in the HTTPRouteRule (and perhaps also in the HTTPRouteMatch) has been biting us in more ways, due to the CEL validations in the HTTPRoute (v0.8+).

[Kuadrant](https://kuadrant.io) policies (based on Policy Attachment, [GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713/)) support associating parts of a policy spec ("policy rules") to specific sections of the main targeted network resource – particularly to HTTPRouteRules, when a HTTPRoute is the target.

Because HTTPRouteRules cannot be referred by name, our solution was to define `matches []HTTPRouteMatch` fields in our policy APIs, which we use to lookup for HTTPRouteRules that declare identical matches, to associate the policy rule portion with. We call these structures "[route selectors](https://docs.kuadrant.io/kuadrant-operator/doc/reference/route-selectors/#routeselector)" (which also include `hostnames`.) A typical Kuadrant policy can contain multiple rules, with multiple route selectors each. See example in a Kuadrant [RateLimitPolicy](https://docs.kuadrant.io/multicluster-gateway-controller/docs/how-to/simple-ratelimitpolicy-for-app-developers/#2-enforce-rate-limiting-on-requests-to-the-toy-store-api) and in a Kuadrant [AuthPolicy](https://docs.kuadrant.io/kuadrant-operator/doc/user-guides/auth-for-app-devs-and-platform-engineers/#3-protect-the-toy-store-application-persona-app-developer).

This would work fine until v0.7, but since validation rules based on [Common Expression Language (CEL)](https://github.com/google/cel-go) have been added to the HTTPRoute (https://github.com/kubernetes-sigs/gateway-api/pull/2253), the estimated cost limit exceeds the thresholds, and we have been struggling with having to set some hard constraints on the number of policy rules a user can declare.

In particular, these HTTPRoute validations have been flagged as the ones adding most to the estimated cost of our APIs: https://github.com/kubernetes-sigs/gateway-api/blame/14197584076cbc6853caf4a6cef29d6d5810595c/apis/v1beta1/httproute_types.go#L193

A much better alternative – that would not only fix the issue of the estimated complexity of the imported validation rules, but also improve assertiveness in how those associations between policy rules and HTTPRouteRules ("HTTPRouteRule refs") are declared – would be having name fields available in the HTTPRouteRule type.

**Comment by costinm** [comment]
For default values: "1", "2"... 

Or match-1 if we want to avoid confusion with array syntax. 

Since most routes have few matchers - just allowing attachment with an array syntax or the default names would work very well with no change to the route.

**Comment by mikemorris** [comment]
This is starting to feel more useful with steps towards standardizing support for partial validity of routes in https://github.com/kubernetes-sigs/gateway-api/pull/2429 and could greatly help clarify corresponding error messages.

Would strongly prefer @youngnick and @maleck13's suggestions that unset optional names be empty/nil and a mismatched sectionName in policy targetRef should be a ResolvedRefs error - array-based indexing would be too easy to accidentally switch targets of a policy when adding or removing a rule from a route.

**Comment by shaneutt** [comment]
@guicassolato you recently discussed your strong need and use case for this, are you interested in being one of the assigned people and to try and move this forward by starting a GEP?

We would ask the first pass at this be focused entirely on the WHAT and WHY and try to stay away from the how (e.g. no API design in the first PR) so we can build consensus on what it is and why we want to do it, but not get lost (yet) in the how we do it.

**Comment by guicassolato** [comment]
@shaneutt, absolutely yes. I think @maleck13 may have brought the topic to the GAMMA meeting today. The two of us will sync up tomorrow and proceed with the next steps to put the GEP together asap. Thank you!

**Comment by shaneutt** [comment]
Awesome! Thank you! Let us know if you get stuck or need help.

/assign @guicassolato

**Comment by guicassolato** [comment]
Opened the draft PR with TLDR, a few goal/non-goals and brief summary of alternatives considered for now. #2593

cc @shaneutt @maleck13

**Comment by mikemorris** [comment]
Can this be closed now that #2985 has been released in v1.2, or is there any more to do here?

**Comment by mikemorris** [comment]
/remove-lifecycle stale

Should this stay open until the feature graduates to standard, or close now?

**Comment by robscott** [comment]
> Should this stay open until the feature graduates to standard, or close now?

We should leave GEP issues open until they graduate to standard, thanks!

**Comment by mikemorris** [comment]
/remove-lifecycle stale

Maybe we can consider moving this from experimental to standard in 1.4?

**Comment by shaneutt** [comment]
This feature has been accepted for the `v1.4.0` release. Please see [this announcement](https://github.com/kubernetes-sigs/gateway-api/discussions/3824) for more details. @guicassolato please create or attach sub-tasks for this issue that provide an overview of all work remaining needed to get this to standard. If you have any questions, concerns, or are in need of support please reach out to the [maintainers](https://github.com/kubernetes-sigs/gateway-api/blob/e7c711b0a91273727288942faa10573164e6981b/OWNERS_ALIASES#L13) so we can assist you!

> **Note**: @mikemorris is a primary reviewer for this feature, however please reach out and see if you can get one more dedicated [reviewer](https://github.com/kubernetes-sigs/gateway-api/blob/e7c711b0a91273727288942faa10573164e6981b/OWNERS_ALIASES#L44) as this will help us keep this moving along smoothly.

**Comment by shaneutt** [comment]
@guicassolato checking in: how are things going? Are you blocked on anything or need any support to help move this forward? Have you been able to find another dedicated review?

**Comment by guicassolato** [comment]
> [@guicassolato](https://github.com/guicassolato) checking in: how are things going? Are you blocked on anything or need any support to help move this forward? Have you been able to find another dedicated review?

PR sent: #3826. Needs review.

**Comment by shaneutt** [comment]
This isn't _quite_ ready to be considered fully complete yet, as noted in https://github.com/kubernetes-sigs/gateway-api/pull/3826#pullrequestreview-2961758398.

We need to track getting some conformance reports prior to `v1.4.0`, so I'm going to add this as a blocker for the release.

@guicassolato please ping anyone you think could be good to provide some feedback on those tests.

**Comment by shaneutt** [comment]
We talked about this one on the meeting today. The status for this appears to be 🟢: The GEP is marked standard, and we have conformance tests. We discussed that there were some runs of the conformance tests that succeeded in Istio, which is a great signal as well. Lior is going to post an update here about that soon.

I marked this as "review" for now to indicate that it's ready for code freeze, API review, and our release candidates.

**Comment by LiorLieberman** [comment]
Confirm I ran it now with istio and it passes these tests.
